pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,64 +15,240 @@
+	if err != nil {
+		klog.Errorf("Error when create aggreagation process: %v", err)
+		return
+	}
+	fag.aggregationProcess = ap
+
+	exportTicker := time.NewTicker(fag.exportFrequency)
 	defer exportTicker.Stop()
-	sentMsgNum := 0
-	go cp.Start()
+	go fag.collectingProcess.Start()
+	go fag.aggregationProcess.Start()
 	for {
 		select {
 		case <-stopCh:
 			break
 		case <-exportTicker.C:
-			curMsgs := cp.GetMessages()
-			klog.Infof("Current message list: %+v\n", curMsgs)
-			for i := sentMsgNum; i < len(curMsgs); i++ {
-				msg := curMsgs[i]
-				setType := msg.Set.GetSetType()
-				templateID := msg.Set.GetRecords()[0].GetTemplateID()
-				setSent := entities.NewSet(setType, templateID, false)
-				for _, record := range msg.Set.GetRecords() {
-					setSent.AddRecord(record.GetInfoElements(), record.GetTemplateID())
+			if fag.exportingProcess == nil {
+				err := fag.initExportingProcess()
+				if err != nil {
+					klog.Errorf("Error when initializing exporting process: %v", err)
+					// There could be other errors while initializing exporting process other than connecting to IPFIX collector,
+					// therefore closing the connection and resetting the process.
+					if fag.exportingProcess != nil {
+						fag.exportingProcess.CloseConnToCollector()
+						fag.exportingProcess = nil
+					}
+					continue
 				}
-				bytesSent, err := ep.AddSetAndSendMsg(setType, setSent)
-				klog.Infof("Msg index: %d, sent bytes: %d, error: %v\n", i, bytesSent, err)
 			}
-			sentMsgNum = len(curMsgs)
+			err := fag.aggregationProcess.ForAllRecordsDo(fag.sendFlowKeyRecords)
+			if err != nil {
+				klog.Errorf("Error when sending flow records: %v", err)
+				// If there is an error when sending flow records because of intermittent connectivity, we reset the connection
+				// to IPFIX collector and retry in the next export cycle to reinitialize the connection and send flow records.
+				fag.exportingProcess.CloseConnToCollector()
+				fag.exportingProcess = nil
+				continue
+			}
+			for _, key := range fag.sentFlowKeys {
+				fag.aggregationProcess.DeleteFlowKeyFromMap(*key)
+			}
+			fag.sentFlowKeys = make([]*ipfixintermediate.FlowKey, 0)
+		}
+	}
+	fag.exportingProcess.CloseConnToCollector()
+	fag.collectingProcess.Stop()
+	fag.aggregationProcess.Stop()
+}
+
+func (fag *flowAggregator) sendFlowKeyRecords(key ipfixintermediate.FlowKey, records []ipfixentities.Record) error {
+	setSent := ipfix.NewSet(ipfixentities.Data, fag.templateID, false)
+	allInfoElements := []string{}
+	allInfoElements = append(IANAInfoElements, IANAReverseInfoElements...)
+	allInfoElements = append(allInfoElements, AntreaInfoElements...)
+	for _, record := range records {
+		elements := make([]*ipfixentities.InfoElementWithValue, 0)
+		for _, ieString := range allInfoElements {
+			element, exist := record.GetInfoElementWithValue(ieString)
+			if !exist {
+				return fmt.Errorf("%s not present.", ieString)
+			}
+			elements = append(elements, element)
+		}
```

Comment:
This seems to be a sanity check, right? 
record already has all the elements with values: record.GetOrderedElementList()
If we can make sure allInfoElements are in the same order as we expect from the record in the intermediate process, we can skip mem alloc in L203 and make it simpler.

Reply:
Yes, the reason I wrote this for loop is because the order of elements from the record in the intermediate process is not as I expect.

Reply:
Could you define the list as per intermediate process? The order of elements in intermediate process does not keep changing, right?